File: test/render.test.ts
import { describe, expect, it } from "vitest";
import { renderJson } from "../src/render/json.js";
import { renderMarkdown } from "../src/render/markdown.js";
import { renderTerminal } from "../src/render/terminal.js";
import type { Verdict } from "../src/types.js";

const flagged: Verdict = {
  stage: "Single small VPS is sufficient (est. $12-20/mo)",
  headroom: "This stack typically serves ~5k daily users at this tier",
  tripwire: "If you add background jobs or exceed ~50GB/mo bandwidth, revisit. Next tier is ~$40/mo.",
  flags: [
    "Found Kubernetes manifests. Adds ~$70/mo and ops burden with no signal you need it yet.",
    "Found a Helm chart. It manages releases across a fleet of services, and this repository holds one.",
  ],
  confidence: "medium",
  confidenceNote: "Confidence: medium. Some of this is inferred from what the repository does not contain.",
  doNothingToday: false,
};

const quiet: Verdict = {
  stage: "Static hosting is sufficient (est. $0-5/mo)",
  headroom: "This stack typically serves ~25k daily visitors at this tier",
  tripwire: "If you add a login, revisit.",
  flags: [],
  confidence: "high",
  confidenceNote: "Confidence: high. The files in this repository point clearly at this answer.",
  doNothingToday: true,
};

describe("terminal", () => {
  it("prints the four fields in order under aligned labels", () => {
    const lines = renderTerminal(flagged).split("\n");
    expect(lines[0]).toBe("Stage:    Single small VPS is sufficient (est. $12-20/mo)");
    expect(lines[1]).toBe("Headroom: This stack typically serves ~5k daily users at this tier");
    expect(lines[2]).toBe("Tripwire: If you add background jobs or exceed ~50GB/mo bandwidth,");
    expect(lines[3]).toBe("          revisit. Next tier is ~$40/mo.");
    expect(lines[4]).toBe("Flags:    Found Kubernetes manifests. Adds ~$70/mo and ops burden with");
    expect(lines[5]).toBe("          no signal you need it yet.");
  });

  it("gives each flag its own wrapped paragraph", () => {
    const lines = renderTerminal(flagged).split("\n");
    expect(lines[6]).toBe("          Found a Helm chart. It manages releases across a fleet of");
    expect(lines[7]).toBe("          services, and this repository holds one.");
  });

  it("never runs past 70 columns", () => {
    for (const line of renderTerminal(flagged).split("\n")) {
      expect(line.length).toBeLessThanOrEqual(70);
    }
  });

  it("says None when nothing is over provisioned, and closes with the verdict", () => {
    const output = renderTerminal(quiet);
    expect(output).toContain("Flags:    None.");
    expect(output.endsWith("\nDo nothing today.\n")).toBe(true);
  });

  it("leaves the closing sentence off when there is something to remove", () => {
    expect(renderTerminal(flagged)).not.toContain("Do nothing today.");
  });

  it("keeps internal metrics out of what a founder reads", () => {
    const output = renderTerminal(flagged) + renderTerminal(quiet);
    expect(output).not.toMatch(/\b(CPU|RPS|p95|p99|IOPS|latency|throughput)\b/i);
  });
});

describe("markdown", () => {
  it("lists flags as bullets and ends with a newline", () => {
    const output = renderMarkdown(flagged);
    expect(output).toContain("**Flags:**\n\n- Found Kubernetes manifests.");
    expect(output.endsWith("\n")).toBe(true);
    expect(output).not.toContain("Do nothing today.");
  });

  it("keeps the four fields in contract order", () => {
    const output = renderMarkdown(quiet);
    expect(output.indexOf("**Stage:**")).toBeLessThan(output.indexOf("**Headroom:**"));
    expect(output.indexOf("**Headroom:**")).toBeLessThan(output.indexOf("**Tripwire:**"));
    expect(output.indexOf("**Tripwire:**")).toBeLessThan(output.indexOf("**Flags:**"));
    expect(output).toContain("**Flags:** None.");
    expect(output).toContain("Do nothing today.");
  });
});

describe("json", () => {
  it("emits the four fields first and in order", () => {
    const parsed = JSON.parse(renderJson(flagged)) as Record<string, unknown>;
    expect(Object.keys(parsed)).toEqual([
      "schema",
      "stage",
      "headroom",
      "tripwire",
      "flags",
      "confidence",
      "confidenceNote",
      "doNothingToday",
    ]);
    expect(parsed["flags"]).toHaveLength(2);
  });
});

describe("every renderer", () => {
  it("reads the same verdict and says the same thing about it", () => {
    const json = JSON.parse(renderJson(quiet)) as Record<string, unknown>;
    for (const output of [renderTerminal(quiet), renderMarkdown(quiet)]) {
      expect(output).toContain(quiet.stage);
      expect(output).toContain(quiet.headroom);
      expect(output).toContain(quiet.tripwire);
      expect(output).toContain("Do nothing today.");
    }
    expect(json["stage"]).toBe(quiet.stage);
    expect(json["doNothingToday"]).toBe(true);
  });
});

File: internal/render/render_test.go
package render

import (
	"encoding/json"
	"regexp"
	"strings"
	"testing"

	"github.com/MithrilBytes/shiftpoint/rules"
)

var flagged = rules.Verdict{
	Stage:    "Single small VPS is sufficient (est. $12-20/mo)",
	Headroom: "This stack typically serves ~5k daily users at this tier",
	Tripwire: "If you add background jobs or exceed ~50GB/mo bandwidth, revisit. Next tier is ~$40/mo.",
	Flags: []string{
		"Found Kubernetes manifests. Adds ~$70/mo and ops burden with no signal you need it yet.",
		"Found a Helm chart. It manages releases across a fleet of services, and this repository holds one.",
	},
	Confidence:     rules.Medium,
	ConfidenceNote: "Confidence: medium. Some of this is inferred from what the repository does not contain.",
	DoNothingToday: false,
}

var quiet = rules.Verdict{
	Stage:          "Free static hosting covers this (est. $0/mo)",
	Headroom:       "Cloudflare Pages serves static files free",
	Tripwire:       "If you add a login, revisit.",
	Confidence:     rules.High,
	ConfidenceNote: "Confidence: high. The files in this repository point clearly at this answer.",
	DoNothingToday: true,
}

func TestTerminalPutsTheFourFieldsInOrderUnderAlignedLabels(t *testing.T) {
	lines := strings.Split(Terminal(flagged), "\n")
	want := []string{
		"Stage:    Single small VPS is sufficient (est. $12-20/mo)",
		"Headroom: This stack typically serves ~5k daily users at this tier",
		"Tripwire: If you add background jobs or exceed ~50GB/mo bandwidth,",
		"          revisit. Next tier is ~$40/mo.",
		"Flags:    Found Kubernetes manifests. Adds ~$70/mo and ops burden with",
		"          no signal you need it yet.",
	}
	for i, expected := range want {
		if lines[i] != expected {
			t.Errorf("line %d\n got: %q\nwant: %q", i, lines[i], expected)
		}
	}
}

func TestTerminalGivesEachFlagItsOwnParagraph(t *testing.T) {
	lines := strings.Split(Terminal(flagged), "\n")
	if lines[6] != "          Found a Helm chart. It manages releases across a fleet of" {
		t.Errorf("second flag not wrapped under the gutter: %q", lines[6])
	}
}

func TestTerminalNeverRunsPastSeventyColumns(t *testing.T) {
	for _, verdict := range []rules.Verdict{flagged, quiet} {
		for _, line := range strings.Split(Terminal(verdict), "\n") {
			if len(line) > width {
				t.Errorf("%d columns: %q", len(line), line)
			}
		}
	}
}

func TestTerminalSaysNoneAndClosesWithTheVerdict(t *testing.T) {
	output := Terminal(quiet)
	if !strings.Contains(output, "Flags:    None.") {
		t.Error("an empty flag list should read None.")
	}
	if !strings.HasSuffix(output, "\nDo nothing today.\n") {
		t.Error("the closing sentence is missing")
	}
	if strings.Contains(Terminal(flagged), "Do nothing today.") {
		t.Error("affirmed no action while flagging something to remove")
	}
}

func TestTerminalKeepsInternalMetricsOut(t *testing.T) {
	metric := regexp.MustCompile(`(?i)\b(CPU|RPS|p95|p99|IOPS|latency|throughput)\b`)
	for _, verdict := range []rules.Verdict{flagged, quiet} {
		if metric.MatchString(Terminal(verdict)) {
			t.Error("an internal metric reached a founder")
		}
	}
}

func TestTerminalCarriesTheConfidenceLine(t *testing.T) {
	// The confidence line is the tool's honesty mechanism and terminal is its
	// primary surface. A renderer that drops it fails silently otherwise.
	for _, verdict := range []rules.Verdict{flagged, quiet} {
		if !strings.Contains(Terminal(verdict), "Confidence: ") {
			t.Error("the confidence line is missing")
		}
	}
}

func TestMarkdownRendersFlagsAsBulletsAndOneAsWell(t *testing.T) {
	many := Markdown(flagged)
	if !strings.Contains(many, "**Flags:**\n\n- Found Kubernetes manifests.") {
		t.Error("multiple flags are not a bullet list")
	}

	// One flag is the common case in real repositories and had no test.
	single := flagged
	single.Flags = flagged.Flags[:1]
	if !strings.Contains(Markdown(single), "**Flags:**\n\n- Found Kubernetes") {
		t.Error("a lone flag is rendered differently from several")
	}

	if !strings.Contains(Markdown(quiet), "**Flags:** None.") {
		t.Error("no flags should read None.")
	}
}

func TestMarkdownKeepsTheFourFieldsInContractOrder(t *testing.T) {
	output := Markdown(quiet)
	order := []string{"**Stage:**", "**Headroom:**", "**Tripwire:**", "**Flags:**"}
	previous := -1
	for _, field := range order {
		at := strings.Index(output, field)
		if at <= previous {
			t.Errorf("%s is out of order", field)
		}
		previous = at
	}
	if !strings.HasSuffix(output, "\n") {
		t.Error("no trailing newline")
	}
}

func TestJSONEmitsTheFourFieldsFirstAndInOrder(t *testing.T) {
	encoded, err := JSON(flagged)
	if err != nil {
		t.Fatal(err)
	}

	var order []string
	decoder := json.NewDecoder(strings.NewReader(encoded))
	if _, err := decoder.Token(); err != nil {
		t.Fatal(err)
	}
	for decoder.More() {
		key, err := decoder.Token()
		if err != nil {
			t.Fatal(err)
		}
		order = append(order, key.(string))
		var discard json.RawMessage
		if err := decoder.Decode(&discard); err != nil {
			t.Fatal(err)
		}
	}

	want := []string{"schema", "stage", "headroom", "tripwire", "flags", "confidence", "confidenceNote", "doNothingToday"}
	if strings.Join(order, ",") != strings.Join(want, ",") {
		t.Errorf("got %v, want %v", order, want)
	}
}

func TestJSONEmitsAnEmptyArrayRatherThanNull(t *testing.T) {
	encoded, err := JSON(quiet)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(encoded, `"flags": []`) {
		t.Errorf("a verdict with no flags should carry an empty array: %s", encoded)
	}
}

func TestEveryRendererReadsTheSameVerdict(t *testing.T) {
	encoded, err := JSON(quiet)
	if err != nil {
		t.Fatal(err)
	}
	for _, output := range []string{Terminal(quiet), Markdown(quiet)} {
		for _, field := range []string{quiet.Stage, quiet.Headroom, quiet.Tripwire} {
			if !strings.Contains(output, field) {
				t.Errorf("a renderer dropped %q", field)
			}
		}
		if !strings.Contains(output, "Do nothing today.") {
			t.Error("a renderer dropped the closing sentence")
		}
	}
	if !strings.Contains(encoded, quiet.Stage) {
		t.Error("json dropped the stage")
	}
}

func TestWrapKeepsALongWordWholeRatherThanCuttingIt(t *testing.T) {
	long := "https://example.com/pricing/compute/instances/general-purpose/details"
	lines := wrap("see "+long+" for more", 30)
	found := false
	for _, line := range lines {
		if line == long {
			found = true
		}
	}
	if !found {
		t.Errorf("a long token was broken: %v", lines)
	}
}
